### PR TITLE
Danfooo bugfixes Bug fix package 8

### PR DESCRIPTION
**Pull request by danfooo**

### CODE REVIEW

File: src/cookie-manager/cookie_manager_views.js [hunk 1]
```diff
@@ -430,6 +430,15 @@ cls.CookieManager.CookieManagerViewBase = function()
       // "domain" is val of [input] (with add_cookie service present), or runtimes .hostname
       var domain       = domain_input && domain_input.value.trim() || runtime && this.data._rts[runtime].hostname;
 
+      // Fix expires value, work around CORE-47780: .value property of <input type=datetime-local>
+      // element has two digits representing milliseconds, instead of three.
+      if (expires.split(".").length > 1)
```
Comment (hzr) [inline src/cookie-manager/cookie_manager_views.js]: Might split only once (instead of this and line 437).

**Comment by hzr** [comment]
Except for the one comment, it looks good.

**Comment by danfooo** [comment]
Okay, also one if statement does it.

**Comment by hzr** [comment]
Looks fine!

File: src/cookie-manager/cookie_manager_views.js [hunk 2]
```diff
@@ -430,6 +430,15 @@ cls.CookieManager.CookieManagerViewBase = function()
       // "domain" is val of [input] (with add_cookie service present), or runtimes .hostname
       var domain       = domain_input && domain_input.value.trim() || runtime && this.data._rts[runtime].hostname;
 
+      // Fix expires value, work around CORE-47780: .value property of <input type=datetime-local>
+      // element has two digits representing milliseconds, instead of three.
+      var expires_parts = expires.split(".");
+      if (expires_parts[1] && expires_parts[1].length < 3)
```
Comment (chriskr) [inline src/cookie-manager/cookie_manager_views.js]: You would get that check a bit cheaper with a reg exp.

Comment (danfooo) [inline src/cookie-manager/cookie_manager_views.js]: Is that just a remark or does it block? :)

Comment (danfooo) [inline src/cookie-manager/cookie_manager_views.js]: For this a case I don't see much benefit in a regular expression. I find it quicker to read as is.

Comment (hzr) [inline src/cookie-manager/cookie_manager_views.js]:
``` javascript
var expires_parts = expires.split(".");
if (expires_parts[1] && expires_parts[1].length < 3)
```

versus e.g.

``` javascript
if (!/\..{3}/.test(expires))
```

I think I prefer the shorter one, but do what you want.

**Comment by danfooo** [comment]
Alright, I was thinking too complicated. It has to be a positive test though, since the .value can skip milliseconds completely. That's when there was no value attribute to begin with and the user sets a date-and-time.

**Comment by hzr** [comment]
Looks fine.

**Comment by danfooo** [comment]
@chriskr @hzr Added one more fix.

**Comment by chriskr** [comment]
Looks good AFAICT.

**Comment by danfooo** [comment]
@hzr yes, when expires is 0 it means the cookie is valid for the session, I forgot that. That cases actually need another fix it seems.

**Comment by danfooo** [comment]
One more fix. Had introduced a regression by dealing with timezone-offset only when reading the value from DOM, not when writing it. Fixed in the latest commit, I hope this is done now.

**Comment by hzr** [comment]
Good!

**Comment by danfooo** [comment]
One more try. @hzr @chriskr

**Comment by danfooo** [comment]
One more bugfix (65b6570) and review fixes. @hzr @chriskr

**Comment by hzr** [comment]
All fine!

File: src/scripts/dom.js [hunk 3]
```diff
@@ -777,6 +777,39 @@ Date.prototype.toLocaleISOString = function()
  return new Date(this.getTime() - this.getTimezoneOffset() * 1000 * 60).toISOString().replace('Z','');
 };
 
+Date.fromLocaleISOString = function(localeISOString)
```
Comment (hzr) [inline src/scripts/dom.js]: I guess this should be on the prototype btw.

**Comment by danfooo** [comment]
See last commit, I moved the workaround for unparsable date-strings out of toLocaleISOString, I think this also makes it clearer where what problem is solved.

I'm against caching anything though, if toLocaleISOString needs to be used at all (or if we suddenly get a UTC string) won't change during a session, but the check is cheap and I'd rather look at the value and decide every time. This is all only used when cookies are edited, not worth storing anything per session.

**Comment by hzr** [comment]
Looks good now.

**Comment by danfooo** [comment]
@hzr @chriskr ^^ one more attempt..

**Comment by hzr** [comment]
Looks fine!

**Comment by chriskr** [comment]
Yes, looks good.

**Comment by p01** [comment]
Goody

**Pull request merged**